pkg/net/libp2p/libp2ptransport.go

```diff
@@ -0,0 +1,283 @@
+	}
+}
+
+func (t *Transport) ImplementsModule() {}
+
+func (t *Transport) EventsOut() <-chan *events.EventList {
+	return t.incomingMessages
+}
+
+func (t *Transport) Start() error {
+	t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s handler starting on %v", t.ownID, t.host.Addrs()))
+	t.host.SetStreamHandler(ID, t.mirHandler)
+	return nil
+}
+
+func (t *Transport) Stop() {
+	t.logger.Log(logging.LevelDebug, "Stopping libp2p transport.")
+	defer t.logger.Log(logging.LevelDebug, "libp2p transport stopped.")
+
+	t.outboundStreamsMx.Lock()
+	defer t.outboundStreamsMx.Unlock()
+
+	for id, s := range t.outboundStreams {
+		if s == nil {
+			continue
+		}
+		t.logger.Log(logging.LevelDebug, "Closing connection", "to", id)
+		if err := s.Close(); err != nil {
+			t.logger.Log(logging.LevelWarn, fmt.Sprintf("Could not close connection to node %v: %v", id, err))
+			continue
+		}
+		t.logger.Log(logging.LevelDebug, "Closed connection", "to", id)
+	}
+
+	if err := t.host.Close(); err != nil {
+		t.logger.Log(logging.LevelError, fmt.Sprintf("Could not close libp2p %v: %v", t.ownID, err))
+	} else {
+		t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %v libp2p host closed", t.ownID))
+	}
+}
+
+func (t *Transport) Connect(ctx context.Context) {
+	wg := sync.WaitGroup{}
+	wg.Add(len(t.membership) - 1)
+
+	for nodeID, nodeAddr := range t.membership {
+		if nodeID == t.ownID {
+			continue
+		}
+
+		go func(nodeID types.NodeID, nodeAddr string) {
+			defer wg.Done()
+
+			t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s is connecting to node %s", t.ownID.Pb(), nodeID.Pb()))
+
+			maddr, err := multiaddr.NewMultiaddr(nodeAddr)
+			if err != nil {
+				t.logger.Log(logging.LevelError,
+					fmt.Sprintf("failed to parse %v for node %v: %v", nodeAddr, nodeID, err))
+				return
+			}
+
+			// Extract the peer ID from the multiaddr.
+			info, err := peer.AddrInfoFromP2pAddr(maddr)
+			if err != nil {
+				t.logger.Log(logging.LevelError,
+					fmt.Sprintf("failed to parse addr %v for node %v: %v", maddr, nodeID, err))
+				return
+			}
+
+			t.host.Peerstore().AddAddrs(info.ID, info.Addrs, peerstore.PermanentAddrTTL)
+
+			s, err := t.openStream(ctx, info.ID)
+			if err != nil {
+				t.logger.Log(logging.LevelError, fmt.Sprintf("couldn't open stream: %v", err))
+				return
+			}
+			t.addOutboundStream(nodeID, s)
+			t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s has connected to node %s", t.ownID.Pb(), nodeID.Pb()))
+
+		}(nodeID, nodeAddr)
+	}
+
+	wg.Wait()
+}
+
+func (t *Transport) openStream(ctx context.Context, p peer.ID) (network.Stream, error) {
+	timeout := 500 * time.Millisecond
+	for {
+		sctx, cancel := context.WithTimeout(ctx, timeout)
+		s, err := t.host.NewStream(sctx, p, ID)
+		cancel()
+
+		if err == nil {
+			return s, nil
+		}
+
+		t.logger.Log(logging.LevelError, "opening stream", "err", err.Error())
+
+		select {
+		case <-time.After(timeout):
+			continue
+		case <-ctx.Done():
+			return nil, fmt.Errorf("context closed")
+		}
```

Comment:
I don't really understand this block... What would happen if we completely removed it?